### PR TITLE
add profiler menu, additional commands and other improvements

### PR DESCRIPTION
**Pull request by javierluraschi**
- Add new profile menu
- Change extension from `rprof` to `Rprof`
- Enable profiling multi-line statements
- Prevent profiling status from being clickable
- Avoid minimizing console after profile creation
- Add support to restore console and restore console on profile start
- Add support to trigger profile help from menu
- Implement save/load profile in profile menu
- Use `Untitled#` for unnamed profiles
- Add support to save unnamed profile to match behavior of other docs

screen shot 2016-03-01 at 4 39 52 pm

### CODE REVIEW

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java [hunk 1]
```diff
@@ -378,6 +375,26 @@ public void revertChanges(Command onCompleted)
    {
       onCompleted.execute();
    }
+   
+   private String getAndSetInitialName()
```
Comment (jjallaire) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java]: I think we should use "Profile1", etc. rather than "Untitled1"

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java [hunk 2]
```diff
@@ -258,6 +253,8 @@ public void execute(String newHtmlPath)
 
    public void onDeactivate()
    {
+      commands_.saveProfileAs().setEnabled(false);
```
Comment (jjallaire) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java]: We should be able to handle this via use of our dynamic commands (commands which are only enabled by file type / active editing target). Here is where dynamic commands are defined (you need to add saveProfileAs to this):

https://github.com/rstudio/rstudio/blob/master/src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java#L280

And here is where you indicate that the profiler file type includes support for this command:

https://github.com/rstudio/rstudio/blob/master/src/gwt/src/org/rstudio/studio/client/common/filetypes/ProfilerType.java#L50-L56

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java [hunk 3]
```diff
@@ -500,8 +531,13 @@ public void onError(ServerError error)
    
    private void persistHtmlPath(String htmlPath)
    {
+      persistDocumentProperty("htmlPath", htmlPath);
```
Comment (jjallaire) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java]: Just wanted to confirm that the html paths are constructed so that they survive from session to session (since the R session temp directory gets blasted at the end of a session). These files probably want to live within the module_context::scopedScratchPath (per-project "state" file storage). @jmcphers Does that sound right to you?

**Comment by jjallaire** [comment]
What do we think about adding the "Save Profile As" command to the toolbar? (perhaps this could be only when the profile is "untitled")

**Comment by jjallaire** [comment]
Let's add the "Publish" button to the far right of the toolbar (@jmcphers can show you how to do this). This solves a big pain point now as the only way to easily publish profile visualizations is to print them to the RStudio Viewer (for which they are too large) then publish from there.

**Comment by jjallaire** [comment]
Let's add the "Profile Selected Line(s)" command to the code menu off of the source editor toolbar. You can do this here: 

https://github.com/rstudio/rstudio/blob/master/src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTargetWidget.java#L441-L469

**Comment by jjallaire** [comment]
Since my comments are more or less for feature additions rather than changes to this code I will go ahead and merge.

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java (hunk 1 above)
Comment (javierluraschi) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java]: I thought about this for a second but noticed that the `nameProvider` is reused by most documents and the convention seems to be `Untitled#` with an icon for the respective document type. However, I do agree `Profiler#` is more clear; will make the change on separate PR.

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java (hunk 2 above)
Comment (javierluraschi) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java]: K. Addressing with subsequent PR.

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java (hunk 3 above)
Comment (javierluraschi) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java]: Yes. we store html files under `/profiler_cache`, similar to `/viewer_Cache` for the `View()` feature.

**Comment by javierluraschi** [comment]
K. Will follow up with improvements in separate PR. Thanks!

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java (hunk 3 above)
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java]: @jjallaire yes, that sounds right; @javierluraschi I think that's the right spot. We could make them (and the viewer cache, for that matter) more scoped since technically these are bound to an RStudio "session", but I don't think it's necessary since the chances of collision are very low.

**Comment by jmcphers** [comment]
@javierluraschi let me know when you want to discuss the publish bits -- it shouldn't be too hard since the publish button is largely self-hosting, and this is almost if not fully analogous to publishing an HTML widget from the viewer pane, but there will be a little bit of glue needed.
